File: strategy/typescript/src/strategy-oop.spec.ts
import { describe, it } from "node:test";

import {
	Notifier,
	EmailNotificationStrategy,
	SMSNotificationStrategy,
	PushNotificationStrategy,
} from "./strategy-oop.js";
import assert from "node:assert";

describe("oop version of strategy pattern", () => {
	it("should notify users via email", async () => {
		const emailStrategy = new EmailNotificationStrategy({
			userEmail: "example@test.com",
			subject: "Important information",
			sendEmail: async ({}) => "success",
		});
		const notifier = new Notifier({
			strategies: { email: { notify: emailStrategy.notify } },
		});

		const result = await notifier.notify({
			provider: "email",
			message: "Notification body",
		});

		assert.equal(result, "success");
	});

	it("should notify users via sms", async () => {
		const smsStrategy = new SMSNotificationStrategy({
			phoneNumber: "+00500500500",
			sendSms: async ({}) => "success",
		});
		const notifier = new Notifier({
			strategies: { sms: { notify: smsStrategy.notify } },
		});

		const result = await notifier.notify({
			provider: "sms",
			message: "Notification content",
		});

		assert.equal(result, "success");
	});

	it("should notify user directly in app via pushNotification", async () => {
		const pushNotificationStrategy = new PushNotificationStrategy({
			userId: "1a",
			sendPushNotification: async ({}) => "success",
		});
		const notifier = new Notifier({
			strategies: {
				pushNotification: { notify: pushNotificationStrategy.notify },
			},
		});

		const result = await notifier.notify({
			provider: "pushNotification",
			message: "Notification content",
		});

		assert.equal(result, "success");
	});
});

File: strategy/python/strategy_oop.spec.py
import unittest

from strategy_oop import EmailNotificationStrategy, Notifier, NotifyParams


class TestNotifier(unittest.IsolatedAsyncioTestCase):
    async def test_email_notification(self):
        async def mock_send_email(params):
            return "success"

        email_strategy = EmailNotificationStrategy(
            params={
                "user_email": "example@test.com",
                "subject": "test subject",
                "send_email": mock_send_email,
            }
        )

        notifier = Notifier({"email": email_strategy})

        result = await notifier.notify(
            NotifyParams(provider="email", message="Notification body")
        )

        self.assertEqual(result, "success")


if __name__ == "__main__":
    unittest.main()
